java/ffi/src/main/java/org/apache/arrow/ffi/FFIReferenceManager.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.ffi;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.OwnershipTransferResult;
+import org.apache.arrow.memory.ReferenceManager;
+import org.apache.arrow.util.Preconditions;
+
+/**
+ * A ReferenceManager implementation that holds a {@link org.apache.arrow.ffi.BaseStruct}.
+ * <p>
+ * A reference count is maintained and once it reaches zero the struct
+ * is released (as per the C data interface specification) and closed.
+ */
+final class FFIReferenceManager implements ReferenceManager {
+  private final AtomicInteger bufRefCnt = new AtomicInteger(0);
+
+  private final BaseStruct struct;
+
+  FFIReferenceManager(BaseStruct struct) {
+    this.struct = struct;
+  }
+
+  @Override
+  public int getRefCount() {
+    return bufRefCnt.get();
+  }
+
+  @Override
+  public boolean release() {
+    return release(1);
+  }
+
+  @Override
+  public boolean release(int decrement) {
+    Preconditions.checkState(decrement >= 1, "ref count decrement should be greater than or equal to 1");
+    // decrement the ref count
+    final int refCnt = bufRefCnt.addAndGet(-decrement);
+    // the new ref count should be >= 0
+    Preconditions.checkState(refCnt >= 0, "RefCnt has gone negative");
+    if (refCnt == 0) {
+      // refcount of this reference manager has dropped to 0
+      // release the underlying memory
+      synchronized (this) {
```

Comment:
maybe we need to check `refCnt == 0` again here?

Reply:
Given that the counter is atomic, in what scenario is it needed? 

I believe that if the counter is decremented to zero then it must be released. Incrementing the counter after (or during) shouldn't change the decision to release. I saw no special code for handling the increment-after-release case in other reference managers so I assume it's not possible by design.

Reply:
Literally speaking `refCnt` is a local variable, it doesn't change.

I think even we check using `bufRefCnt.get()`, that is still not enough. We don't lock writes from other places so the value might happen to be written from 0 to 1 to 0 during this block. That is an ABA problem.

Reply:
@roee88 `BufferLedger` has a check at line 185 and it seems that we don't have it here

https://github.com/apache/arrow/blob/3bbec3f31ca16dd347b9ec88bd2b38fbf1a403b9/java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferLedger.java#L178-L186

Since `release(int decrement)` and `retain(int increment)` are both public so I think one could call retain after calling release manually.

Reply:
Thanks. I'll try to mimic the buffer ledger implementation although I'm not sure how to handle the first retain yet.

@zhztheplayer can you please have a quick look at the OrcReferenceManager implementation to see if a ticket needs to be open for a fix there.

Reply:
Why would a reference count go from 0 to 1? Once it drops to 0, it should never increase again.

Reply:
I agree that in ORC we have similar problems (maybe tickets needed), but I think ORC's case is less severe: it internally uses an unordered_map to store the pointer payload so releasing already-released buffer from Java side might not lead to UB on native side.

https://github.com/apache/arrow/blob/master/cpp/src/jni/orc/concurrent_map.h#L22

But it's likely that in C data interface we can have more complicated cases so there might be possible that a Java-side release finally causes a native double-free or SIGABRT or something, if the release hook doesn't include a check.

Reply:
> Why would a reference count go from 0 to 1? Once it drops to 0, it should never increase again.

It's the pain of manual reference management. A buffer can be released from one place and at the same time be retained from other places. Especially when the life cycle of the managed objects is not designed and gone through carefully by developers.

Reply:
That sounds like a bug to me. Again, a reference count dropped to 0 means that _nobody_ has a strong reference anymore. It can't reappear again.

Reply:
To put it differently: once the reference count drops to zero, the underlying object is released. Doing anything else with it will result in UB (probably crashes or accessing unrelated memory).

Reply:
Yes they are bugs (this type of bugs keep appearing when manual reference management is used in large complicated system). Adding such checks helps debugging rather than dealing with UBs.

Reply:
Also we may be able to move all the checks to a debug option? I am not sure.

Reply:
If by "check" you mean it writes a nice error message, then that sounds reasonable indeed.

Reply:
Yes. Error message is not created here. We may need some.

Reply:
I'll add a message. Can you check if the current implementation is correct? I think that we can omit the synchronized block.

Reply:
It now looks correct to me. We don't seem to need the synchronization block so long as the block will be entered only once. Although in my opinion people can catch exceptions from `#retain` by their own. That's a rare case so we either remove the sync block or expand it to the same scope as in 'BufferLedger'.